CMakeLists.txt

```diff
@@ -126,6 +126,26 @@ option(IREE_ENABLE_LLD "Use lld when linking" OFF)
 option(IREE_ENABLE_ASAN "Enable address sanitizer" OFF)
 option(IREE_ENABLE_MSAN "Enable memory sanitizer" OFF)
 option(IREE_ENABLE_TSAN "Enable thread sanitizer" OFF)
+option(IREE_BYTECODE_MODULE_ENABLE_TSAN "Enable thread sanitizer in IREE modules in tests" OFF)
+
+if(NOT IREE_ENABLE_TSAN STREQUAL IREE_BYTECODE_MODULE_ENABLE_TSAN)
```

Comment:
Branching from other thread

>> gcmn:
>> I'm also a bit worried about `STREQUAL` here since in stringly-typed CMake there are many strings that mean "true" (`1`, `ON`, `YES`, `TRUE`, `Y`, or a non-zero number) "false" (`0`, `OFF`, `NO`, `FALSE`, `N`, `IGNORE`, `NOTFOUND`, the empty string, or ends in the suffix `-NOTFOUND`) :nauseated_face: (from https://cmake.org/cmake/help/latest/command/if.html#basic-expressions). Maybe we really do want to be stricter here, but something to consider

> bjacob:
> do you mean I should be able to use `EQUAL` here ?  I dont really understand that that does, sorry. Is it "equality comparison after implict conversion to int/bool" ?

No `EQUAL` requires that the operands be numbers. When you do `if(FOO)` CMake evaluates `FOO` for a bunch of potentially truth-valued strings, not just e.g. `ON`. So if someone sets `IREE_ENABLE_TSAN=Y` and `IREE_BYTECODE_MODULE_ENABLE_TSAN=1` those are both technically true but `if(NOT IREE_ENABLE_TSAN STREQUAL IREE_BYTECODE_MODULE_ENABLE_TSAN)` will evaluate true (the error condition). The documentation for CMake `option` only talks about `ON`/`OFF` (https://cmake.org/cmake/help/v3.20/command/option.html), but AFAICT that's not actually enforced in any way (I just tried setting an option to `FOO` and it worked fine). I think the only way to force the boolean conversion would be to use `if` itself. Something like

```cmake
if(IREE_ENABLE_TSAN)
  set(IREE_ENABLE_TSAN ON)
else()
  set(IREE_ENABLE_TSAN OFF)
endif()

if(IREE_BYTECODE_MODULE_ENABLE_TSAN)
  set(IREE_BYTECODE_MODULE_ENABLE_TSAN ON)
else()
  set(IREE_BYTECODE_MODULE_ENABLE_TSAN OFF)
endif()
```

or

```cmake
if ((IREE_ENABLE_TSAN AND NOT IREE_BYTECODE_MODULE_ENABLE_TSAN) OR (IREE_BYTECODE_MODULE_ENABLE_TSAN AND NOT IREE_ENABLE_TSAN))
  # error
endif()
```

I'm not sure it's really worth it, but something to note. Ugh CMake :vomiting_face:

Reply:
OK, thanks for explaining. It's what I had feared -- and why I stuck to `STREQUAL`. I understand the concern about overly relying on the particular string truth-values, but like you say, the alternatives are not ideal either. Doing

```
if(IREE_ENABLE_TSAN)
  set(IREE_ENABLE_TSAN ON)
else()
  set(IREE_ENABLE_TSAN OFF)
endif()
```

for every option would be verbose. Doing it only for the few options that we happen to rely on that for at the moment, would be brittle.

Doing

```
if ((IREE_ENABLE_TSAN AND NOT IREE_BYTECODE_MODULE_ENABLE_TSAN) OR (IREE_BYTECODE_MODULE_ENABLE_TSAN AND NOT IREE_ENABLE_TSAN))
  # error
endif()
```

would be verbose and easy for logic bugs to sneak into.  `STREQUAL` still looks reasonable by comparison...

Reply:
Yeah, more just something to be aware of. I agree that if we were to do option 1, we should really do it for all our options. Can you just add a comment here about that deliberate tradeoff to stick to the simpler STREQUAL even though it doesn't quite match with CMake's absurdity

Reply:
Done